tfsdk/schema.go

```diff
@@ -84,6 +84,10 @@ func (s Schema) AttributeTypeAtPath(path *tftypes.AttributePath) (attr.Type, err
 		return n.AttributeType(), nil
 	}
 
+	if s, ok := rawType.(Schema); ok {
```

Comment:
If you're doing this because you got an error when walking the path, I think you're just going to want to return from the transform early if the attributepath is empty, instead. 🤔

Reply:
What would then be the expected behavior? I'm not sure how return early (no transform) helps us, but if I'm missing something please do explain!

If I'm understanding correctly, not returning the root schema attributes at "no" path would not work with this `ImportState` code:

https://github.com/hashicorp/terraform-plugin-framework/blob/f8cc3a2692a533fcbe8aee74e942f2cf11f0d35b/tfsdk/serve_import.go#L88-L99

And testing:

https://github.com/hashicorp/terraform-plugin-framework/blob/f8cc3a2692a533fcbe8aee74e942f2cf11f0d35b/tfsdk/serve_import_test.go#L63-L108

To ensure that actually works (e.g. one of the reasons we are making these changes), the code would then need to be updated here in `setAttributeTransformFunc()` from:

```go
parentPath := path.WithoutLastStep()
parentAttrType, err := s.Schema.AttributeTypeAtPath(parentPath)
```

To something like:

```go
parentPath := path.WithoutLastStep()

var err error
var parentAttrType attr.Type
if len(parentPath.Steps()) == 0 {
	parentAttrType = s.Schema.AttributeType
} else {
	parentAttrType, err = s.Schema.AttributeTypeAtPath(parentPath)
}
```

Although it seems that any future code that potentially hits the "no path" situation would also need to implement the same logic each time.

Reply:
So I think what I'm trying to say is my mental model of the responsibilities of this code are:

1. `Schema.AttributeType` should be used to get the `attr.Type` of the entire schema.
2. `Schema.AttributeTypeAtPath` should be used to get the `attr.Type` of a specific attribute.

I don't know that I consider it a big deal in isolation, but I _am_ slightly concerned about the conflation of the two when used with the transform function.

Now that I've typed this all out I have realised I had confused myself. I was concerned that we're unnecessarily modifying the _entire resource's_ value when we only need to modify the value for a single attribute. Now that I am looking at it again, and started typing out that objection above, I realise we _do_ need to modify the entire resource's value, because transforms operate on the entire resource, not on a specific attribute of a resource. Because the specific attribute of a resource may have been created during the transform, and so the root needs to be modifiable to be updated to include it. I think. Now I'm unsure and think we _could_ get away with just skipping this in the transform, because it would replace the specific attribute.

This has been a journey, y'all. But I think I'm still on team "an error here probably makes sense, I don't think the transform function should _need_ the root schema's `attr.Type` because it shouldn't be modifying it at all." Unless the root schema is `null`.

....

... which is the exact situation we're trying to support. OK. I'm back on "I see why we're doing this", carry on.

I'd like off this rollercoaster please.